design/ceph/admission-controller.md

```diff
@@ -0,0 +1,148 @@
+
+#Created after CSR is approved
+openssl base64 -d -A -out ${PUBLIC_KEY_NAME}.pem 
+```
+
+* Creating and sending a CSR to kubernetes for approval
+
+**Sample : Certificate Signing Request (CSR) in Kubernetes**
+```
+apiVersion: certificates.k8s.io/v1beta1
+kind: CertificateSigningRequest
+metadata:
+  name: ${csrName}
+spec:
+  request: $(cat server.csr | base64 | tr -d '\n')
+  usages:
+  - digital signature
+  - key encipherment
+  - server auth
+```
+Verify with
+` kubectl get csr ${csrName}`
+
+**Sample : Approval of Signed Certificate**
+
+```
+kubectl certificate approve ${csrName}
+```
+
+If it is approved, then we can check the certificate with following command 
+
+```
+$(kubectl get csr ${csrName} -o jsonpath='{.status.certificate}')
+```
+
+* Once approved, a generic secret will be created with the given public and private key which will be later mounted onto the server pod for use.
+
+**Sample : Creating a Secret in Kubernetes**
+```
+kubectl create secret generic ${secret} \
+        --from-file=key.pem=${PRIVATE_KEY_NAME}.pem \
+        --from-file=cert.pem=${PUBLIC_KEY_NAME}.pem
+```
+* Modifying the webhook config to inject CA bundle onto the ValidatingWebhookConfig
+* All the above resources will be created in rook-ceph namespace
+
+
+Using the above approach, the dev/admins will have the responsibility of rotating the certificates when they expire.
+
+Below is an excerpt of what a ValidatingWebhookConfig looks like 
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: demo-webhook
+webhooks:
+  - name: webhook-server.webhook-demo.svc
+    clientConfig:
+      service:
+        name: webhook-server
+        namespace: rook-ceph
+        path: "/validate"
+      caBundle: ${CA_PEM_B64}
+    rules:
+      - operations: [ "CREATE" ]
+        apiGroups: [""]
+        apiVersions: ["v1"]
+        resources: ["pods"]
+```
+
+We can make changes to the above values according to intercept based on whether a resource is being updated/deleted/created or change the type of resource or the request path which will be sent to the server.
+
+## Setup in Rook
+
+### Starting the Webhook Server
+
+ Based on the whether the  secrets are present, the rook operator will  deploy the relevant configuration files onto the cluster and start the server. 
+
+The secrets will be volume mounted on the rook operator pod dynamically when they are detected. After the volumes are mounted, an HTTP server would be started.
+
+Once the server starts, it will look for the appropriate tls key and crt files in the mounted volumes and start intercepting requests based on the path set in ValidatingWebhookConfig.
+
+### Error Handling
+
+If the server is unable to find the proper secrets certificates, It will simply not deploy any admission controller components onto the cluster and hence rook will continue to function normally as before.
```

Comment:
```suggestion
If the server is unable to find valid certificates, it will not deploy any admission controller components onto the cluster and hence rook will continue to function normally as before.
```

Reply:
Fixed